flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/StreamTaskNetworkInputTest.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.runtime.io;
+
+import org.apache.flink.api.common.ExecutionConfig;
+import org.apache.flink.api.common.typeinfo.BasicTypeInfo;
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.apache.flink.runtime.io.disk.iomanager.IOManagerAsync;
+import org.apache.flink.runtime.io.network.api.EndOfPartitionEvent;
+import org.apache.flink.runtime.io.network.api.serialization.RecordDeserializer;
+import org.apache.flink.runtime.io.network.partition.consumer.StreamTestSingleInputGate;
+
+import org.junit.Test;
+import org.powermock.api.mockito.PowerMockito;
+import org.powermock.core.classloader.annotations.PrepareForTest;
+
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
+
+/**
+ * Tests for {@link StreamTaskNetworkInput}.
+ */
+@PrepareForTest(StreamTaskNetworkInput.class)
+public class StreamTaskNetworkInputTest {
+
+	@Test
+	public void testReleasingDeserializerTimely()
+		throws Exception {
+
+		int numInputChannels = 2;
+
+		TypeSerializer<String> inSerializer = BasicTypeInfo.STRING_TYPE_INFO.createSerializer(new ExecutionConfig());
+		StreamTestSingleInputGate inputGate = new StreamTestSingleInputGate<>(numInputChannels, 1024, inSerializer);
+
+		StreamTaskNetworkInput networkInput = new StreamTaskNetworkInput(
+			new BarrierBuffer(inputGate.getInputGate(), new CachedBufferBlocker(512)),
+			inSerializer,
+			new IOManagerAsync(),
+			0);
+
+		RecordDeserializer<?>[] recordDeserializers = (RecordDeserializer<?>[]) PowerMockito
```

Comment:
instead of using mockito to check for private implementation details, you can inject test no-op/wrapping implementations `recordDeserializers`, `checkpointedInputGate` and check whether respected cleaning methods were called.

Relaying on private field access is fragile and basically just duplicates the production code.